nixos/modules/system/activation/top-level.nix

```diff
@@ -298,8 +308,22 @@ in
         config.system.copySystemConfiguration
         ''ln -s '${import ../../../lib/from-env.nix "NIXOS_CONFIG" <nixos-config>}' \
             "$out/configuration.nix"
+        '' +
+      optionalString
+        (config.system.forbiddenDependenciesRegex != "")
+        ''
+          if devPaths="$(grep -E -- "$forbiddenDependenciesRegex" $closureInfo/store-paths)"; then
+            printf "System closure '%s' contains the following disallowed paths:\n%s" "$toplevel" "$devPaths"
+            exit 1
+          fi
         '';
 
+    system.systemBuilderArgs = lib.optionalAttrs (config.system.forbiddenDependenciesRegex != "") {
+      inherit (config.system) forbiddenDependenciesRegex;
+      # can't use config.system.build.toplevel due to infinite recursion(and to allow using extraDependencies to add forbidden dependencies)
+      closureInfo = pkgs.closureInfo { rootPaths = [ config.system.path config.system.build.etc ]; };
```

Comment:
Did you try something like `toplevel.overrideAttrs (o: { extraDependencies = []; closureInfo = null; })`? (if that makes sense)

Reply:
switched to that